modules/javafx.controls/src/test/java/test/javafx/scene/control/TextInputControlTest.java

```diff
@@ -1865,6 +1886,64 @@ public void caretAndAnchorPositionAfterSettingText() {
         assertEquals("", textInput.getText());
     }
 
+    @Test public void test_redo_replaceText_selectionShortening() {
+        textInput.setText("0123456789");
+        assertEquals("0123456789", textInput.getText());
+
+        textInput.replaceText(8, 10, "x");
+        assertEquals("01234567x", textInput.getText());
+
+        textInput.undo();
+        assertEquals("0123456789", textInput.getText());
+
+        textInput.redo();
+        assertEquals("01234567x", textInput.getText());
+    }
+
+    @Test public void replaceSelectionAtEndWithListener() {
+        StringBuilder selectedTextLog = new StringBuilder();
+        StringBuilder selectionLog = new StringBuilder();
+        textInput.setText("x xxx");
+        textInput.selectRange(2, 5);
+        textInput.selectedTextProperty().addListener((__, ___, selection) -> selectedTextLog.append("|" + selection));
+        textInput.selectionProperty().addListener((__, ___, selection) -> selectionLog.append("|" + selection.getStart() + "," + selection.getEnd()));
+        textInput.replaceSelection("a");
+        assertEquals("|", selectedTextLog.toString());
+        assertEquals("|3,3", selectionLog.toString());
+        assertEquals("x a", textInput.getText());
+    }
+
+    @Test public void testSelectionProperties() {
+        textInput.setText("abcdefghij");
+
+        StringBuilder selectedTextLog = new StringBuilder();
+        StringBuilder selectionLog = new StringBuilder();
+        StringBuilder textLog = new StringBuilder();
+        textInput.selectedTextProperty().addListener((__, ___, selection) -> selectedTextLog.append("|" + selection));
+        textInput.selectionProperty().addListener((__, ___, selection) -> selectionLog.append("|" + selection.getStart() + "," + selection.getEnd()));
+        textInput.textProperty().addListener((__, ___, text) -> textLog.append("|" + text));
```

Comment:
The pattern of using multiple underscores for unused variables isn't one we use. I recommend using something more common like `obs` and `o` (or similar).

Reply:
Replaced the variable names in next commit. It looks like "observable" / "oldValue" / "newValue" are almost exclusively used in the current code base so I changed them to that.